sql/src/main/java/io/crate/statistics/MostCommonValues.java

```diff
@@ -0,0 +1,250 @@
+            numMcv = numTracked;
+        } else {
+            if (numMcv > numTracked) {
+                numMcv = numTracked;
+            }
+            if (numMcv > 0) {
+                int[] mcvCounts = new int[numMcv];
+                for (int i = 0; i < numMcv; i++) {
+                    mcvCounts[i] = candidates[i].count;
+                }
+                numMcv = decodeHowManyCommonValuesToKeep(
+                    mcvCounts, numMcv, approxDistinct, nullFraction, samples.size(), numTotalRows);
+            }
+        }
+        if (numMcv == 0) {
+            return EMPTY;
+        }
+        Object[] values = new Object[numMcv];
+        double[] frequencies = new double[numMcv];
+        for (int i = 0; i < numMcv; i++) {
+            values[i] = samples.get(candidates[i].first);
+            frequencies[i] = (double) candidates[i].count / (double) samples.size();
+        }
+        return new MostCommonValues(values, frequencies);
+    }
+
+    private static int decodeHowManyCommonValuesToKeep(int[] mcvCounts,
+                                                       int numMcv,
+                                                       double approxDistinct,
+                                                       double nullFraction,
+                                                       int numSampleRows,
+                                                       long numTotalRows) {
+        assert mcvCounts.length == numMcv : "mcvCounts.length must be equal to numMcv";
+        if (numSampleRows == numTotalRows || numTotalRows <= 1) {
+            // Entire table was sampled, keep all candidates
+            return numMcv;
+        }
+        /* From PostgreSQL:
+         *
+         * Exclude the least common values from the MCV list, if they are not
+         * significantly more common than the estimated selectivity they would
+         * have if they weren't in the list.  All non-MCV values are assumed to be
+         * equally common, after taking into account the frequencies of all the
+         * values in the MCV list and the number of nulls (c.f. eqsel()).
+         *
+         * Here sumcount tracks the total count of all but the last (least common)
+         * value in the MCV list, allowing us to determine the effect of excluding
+         * that value from the list.
+         *
+         * Note that we deliberately do this by removing values from the full
+         * list, rather than starting with an empty list and adding values,
+         * because the latter approach can fail to add any values if all the most
+         * common values have around the same frequency and make up the majority
+         * of the table, so that the overall average frequency of all values is
+         * roughly the same as that of the common values.  This would lead to any
+         * uncommon values being significantly overestimated.
+         */
+        int sumCount = 0;
+        for (int i = 0; i < numMcv - 1; i++) {
```

Comment:
`i < numMcv - 1` looks like the last item in `mcvCounts` is not accounted. Do I miss something here?

Reply:
This is intentionally, see the comment:

>  Here sumcount tracks the total count of all but the last (least common)
> * value in the MCV list, allowing us to determine the effect of excluding
> * that value from the list.

Should I rename the variable to `sumCountExclLast` or something like that?

Reply:
Ah right, sry missed that. It's fine for me as it is.